File: cpp/dicLogic.cpp
#include "dicLogic.hpp"
#include "utils.hpp"
#include <algorithm>
#include <regex>

bool fitsFilter(const std::string &item, const std::string &charFilter)
{
    return item.find_first_not_of(charFilter) == std::string::npos;
}

bool fitsFilterCharCounts(const std::string &item, const std::string &charFilter, const std::vector<size_t> &charCounts)
{
    if (!fitsFilter(item, charFilter))
    {
        return false;
    }

    for (size_t i = 0; i < charFilter.length(); i++)
    {
        ptrdiff_t filterCharCount = std::count(item.begin(), item.end(), charFilter.at(i));
        if (filterCharCount > 0 && (size_t)filterCharCount > charCounts.at(i))
        {
            return false;
        }
    }

    return true;
}

WordList::WordList(std::vector<std::string> &rawWordList, Rules &affixRules, const std::string &charFilter, const std::vector<size_t> &charCounts, std::vector<std::string> &blackList)
{
    for (const std::string &line : rawWordList)
    {
        if (trim(line).length() > 0 && line.at(1) != '#' && !strIsDigit(line))
        {
            std::vector<std::string> elements = split(line, "/");
            std::string baseWord = lowerCase(elements.at(0));

            std::vector<std::string> baseList;
            if (!stringVectorContains(blackList, baseWord) && pushToList(charFilter, charCounts, baseWord) && elements.size() > 1)
            {
                baseList.clear();

                for (const char &flag : elements.at(1))
                {
                    Rule *rule = affixRules.findRule(flag);

                    if (rule)
                    {
                        for (const RuleSet *ruleSet : *rule->getRuleSets())
                        {
                            if (!ruleSet->substitution.empty())
                            {
                                std::regex regEx = std::regex(ruleSet->condition, std::regex_constants::icase);

                                if (std::regex_search(baseWord, regEx, std::regex_constants::match_not_null))
                                {
                                    regEx = std::regex(ruleSet->strippingChars, std::regex_constants::icase);
                                    std::string wordForm = std::regex_replace(baseWord, regEx, ruleSet->substitution);

                                    if (!stringVectorContains(baseList, wordForm))
                                    {
                                        baseList.push_back(wordForm);
                                        pushToList(charFilter, charCounts, wordForm);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    cleanDoubles();
}

bool WordList::pushToList(const std::string &charFilter, const std::vector<size_t> &charCounts, const std::string &item)
{
    if (!item.empty() && fitsFilterCharCounts(item, charFilter, charCounts))
    {
        wordList.push_back(item);
        return true;
    }

    return false;
}

void WordList::cleanDoubles()
{
    std::sort(wordList.begin(), wordList.end());

    for (size_t i = wordList.size() - 1; i > 0; i--)
    {
        if (wordList.at(i) == wordList.at(i - 1))
        {
            wordList.erase(wordList.begin() + i);
        }
    }
}

bool compareByLengthAsc(const std::string &string1, const std::string &string2)
{
    return string1.length() < string2.length();
}

void WordList::prepareWordList(const size_t &minWordLength)
{
    std::sort(wordList.begin(), wordList.end(), compareByLengthAsc);

    auto isMinWordLength = [&](std::string string)
    {
        return string.length() >= minWordLength;
    };

    wordList.erase(wordList.begin(), std::find_if(wordList.begin(), wordList.end(), isMinWordLength));
}

File: C#/dicLogic.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace LetterPuzzle
{
  public static class Filter
  {
    private static Boolean FitsFilter(string item, string charFilter)
    {
      foreach (char itemChar in item)
      {
        Boolean found = false;

        foreach (char filterChar in charFilter)
        {
          if (itemChar == filterChar)
          {
            found = true;
            break;
          }
        }

        if (!found)
        {
          return false;
        }
      }

      return true;
    }

    public static Boolean FitsFilterCharCounts(string item, string charFilter, List<int> charCounts)
    {
      if (!FitsFilter(item, charFilter))
      {
        return false;
      }

      for (int i = 0; i < charFilter.Length; i++)
      {

        int filterCharCount = 0;

        foreach (char itemChar in item)
        {
          if (itemChar == charFilter[i])
          {
            filterCharCount++;
          }
        }

        if (filterCharCount > charCounts[i])
        {
          return false;
        }
      }

      return true;
    }
  };

  class WordList
  {
    private Boolean StrIsDigit(string str)
    {
      return !string.IsNullOrEmpty(str) && str.All(char.IsDigit);
    }

    public WordList(string[] rawWordList, Rules affixRules, string charFilter, List<int> charCounts, List<string> blackList)
    {
      List = new List<string>();

      foreach (string line in rawWordList)
      {
        if (line.Trim().Length > 0 && line[0] != '#' && !StrIsDigit(line))
        {
          string[] elements = line.Split('/');
          string baseWord = elements[0].ToLower();

          List<string> baseList = new List<string>();
          if (!blackList.Contains(baseWord) && PushToList(charFilter, charCounts, baseWord) && elements.Length > 1)
          {
            baseList.Clear();

            foreach (char flag in elements[1])
            {
              Rule rule = affixRules.FindRule(flag);

              if (rule != null)
              {
                foreach (RuleSet ruleSet in rule.RuleSets)
                {
                  if (!string.IsNullOrEmpty(ruleSet.Substitution))
                  {
                    Regex regExp = new Regex(ruleSet.Condition, RegexOptions.IgnoreCase);

                    if (regExp.IsMatch(baseWord))
                    {
                      regExp = new Regex(ruleSet.StrippingChars, RegexOptions.IgnoreCase);
                      string wordForm = regExp.Replace(baseWord, ruleSet.Substitution);

                      if (!List.Contains(wordForm))
                      {
                        baseList.Add(wordForm);
                        PushToList(charFilter, charCounts, wordForm);
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }

      CleanDoubles();
    }

    public Boolean PushToList(string charFilter, List<int> charCounts, string item)
    {
      if (!string.IsNullOrEmpty(item) && Filter.FitsFilterCharCounts(item, charFilter, charCounts))
      {
        List.Add(item);
        return true;
      }

      return false;
    }

    public void CleanDoubles()
    {
      List.Sort();

      for (int i = List.Count() - 2; i >= 0; i--)
      {
        if (List[i] == List[i + 1])
        {
          List.RemoveAt(i + 1);
        }
      }
    }

    private static int CompareByLengthAsc(string string1, string string2)
    {
      return string1.Length - string2.Length;
    }

    public void PrepareWordList(int minWordLength)
    {
      List.Sort(CompareByLengthAsc);

      int i = 0;
      while (i < List.Count && List[i].Length < minWordLength)
      {
        i++;
      }

      List.RemoveRange(0, i);
    }
    public List<string> List { get; private set; }

  };
}
